src/cultionet/losses/losses.py

```diff
@@ -168,27 +322,52 @@ def forward(
         Returns:
             Tanimoto distance loss (float)
         """
-        if inputs.shape[1] > 1:
-            inputs, targets = self.preprocessor(inputs, targets)
+        if self.transform_logits:
+            if len(inputs.shape) == 1:
+                inputs, __ = self.preprocessor(inputs)
+            else:
+                if inputs.shape[1] == 1:
+                    inputs, __ = self.preprocessor(inputs)
+                else:
+                    inputs, targets = self.preprocessor(inputs, targets)
+        else:
+            if len(inputs.shape) > 1:
+                if inputs.shape[1] > 1:
+                    targets = one_hot(targets, dims=inputs.shape[1])
 
         if len(inputs.shape) == 1:
             inputs = inputs.unsqueeze(1)
         if len(targets.shape) == 1:
             targets = targets.unsqueeze(1)
 
         def tanimoto_loss(yhat: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+            y = y.to(dtype=yhat.dtype)
+            if self.scale_pos_weight:
+                if self.class_counts is None:
+                    class_counts = y.sum(dim=0)
+                else:
+                    class_counts = self.class_counts
+                effective_num = 1.0 - self.beta ** class_counts
+                weights = (1.0 - self.beta) / effective_num
+                weights = weights / weights.sum() * class_counts.shape[0]
+            else:
+                weights = torch.ones(
+                    inputs.shape[1], dtype=inputs.dtype, device=inputs.device
+                )
+            # Reduce
             tpl = (yhat * y).sum(dim=0)
             sq_sum = (yhat**2 + y**2).sum(dim=0)
-            numerator = tpl + self.smooth
-            denominator = (sq_sum - tpl) + self.smooth
+            numerator = tpl * weights + self.smooth
```

Comment:
We added optional class weights, which are derived from each batch and designed to balance classes (i.e., more weight is given to under-represented classes).